PyInstaller/hooks/hook-matplotlib.backends.py

```diff
@@ -10,71 +10,125 @@
+from PyInstaller import isolated
 
 
-def get_matplotlib_backend_module_names():
+@isolated.call
+def _default_mpl_backend():
     """
-    List the names of all matplotlib backend modules importable under the current Python installation.
+    Returns the name of the default matplotlib backend.
 
-    Returns
-    ----------
-    list
-        List of the fully-qualified names of all such modules.
+    The default backend is controlled by rcparams and/or MPLBACKEND environment variable.
     """
-    # Statement safely importing a single backend module.
-    import_statement = """
-import os, sys
-
-# Preserve stdout.
-sys_stdout = sys.stdout
-
-try:
-    # Redirect output printed by this importation to "/dev/null", preventing such output from being erroneously
-    # interpreted as an error.
-    with open(os.devnull, 'w') as dev_null:
-        sys.stdout = dev_null
-        __import__('%s')
-# If this is an ImportError, print this exception's message without a traceback. ImportError messages are human-readable
-# and require no additional context.
-except ImportError as exc:
-    sys.stdout = sys_stdout
-    print(exc)
-# Else, print this exception preceded by a traceback. traceback.print_exc() prints to stderr rather than stdout and must
-# not be called here!
-except Exception:
-    sys.stdout = sys_stdout
-    import traceback
-    print(traceback.format_exc())
-"""
-
-    # List of the human-readable names of all available backends.
-    backend_names = eval_statement('import matplotlib; print(matplotlib.rcsetup.all_backends)')
+    import matplotlib
+    return matplotlib.get_backend()
 
+
+@isolated.call
+def _available_mpl_backends():
+    """
+    Returns the names of all available matplotlib backends.
+    """
+    import matplotlib
+    return matplotlib.rcsetup.all_backends
+
+
+@isolated.decorate
+def _check_mpl_backend_importable(module_name):
+    """
+    Attempts to import the given module name (matplotlib backend module).
+
+    Exceptions are propagated to caller.
+    """
+    __import__(module_name)
+
+
+def _recursive_scan_code_objects_for_mpl_use(co):
+    from PyInstaller.depend.bytecode import any_alias, recursive_function_calls
+
+    mpl_use_names = {
+        *any_alias("matplotlib.use"),
+        *any_alias("mpl.use"),  # matplotlib is commonly aliased as mpl
+    }
+
+    backends = []
+    for calls in recursive_function_calls(co).values():
+        for name, args in calls:
+            if not len(args) == 1 or not isinstance(args[0], str):
+                continue
+            if name in mpl_use_names:
+                backends.append(args[0])
+
+    return backends
+
+
+def _autodetect_used_backends(hook_api):
+    # Scan the code for matplotlib.use()
+    modulegraph = hook_api.analysis.graph
+    mpl_code_objs = modulegraph.get_code_using("matplotlib")
+    used_backends = []
+    for name, co in mpl_code_objs.items():
+        used_backends += _recursive_scan_code_objects_for_mpl_use(co)
+
```

Comment:
This block of code appears to be breeding. I really need to try and generalise it into a its own method.

Reply:
I suspect there's not much room left for further abstraction; I just noticed that [`matplotlib.use()`](https://matplotlib.org/stable/api/matplotlib_configuration_api.html#matplotlib.use) can actually be called with two arguments, which means the argument validation will need to be a bit different from other bytecode-scanning blocks we have so far.

On a somewhat related note; is there a way to automatically deal with import aliasing (i.e., can we automatically determine that `mpl.use` actually corresponds to `matplotlib.use`?). And from a slightly different angle: does bytecode analysis allow us to determine what the given symbol is actually bound to? I.e., if we encounter a call to `use('something')`, do we have means to check that this `use` is actually `matplotlib.use` and not coming from some other module (or current scope)?

Reply:
You mean detect what import style was used?
* `from matplotlib import use` -> search for `use()`
* `import matplotlib` -> search for `matplotlib.use()`
Either we'd need to see if modulegraph provides this information or devise our own import detection. The latter option could be a first step towards eventually replacing modulegraph completely which I'd be all for but it would take a lot of work.

Reply:
I was thinking more along the lines of handling different aliases for imported module. I.e., we now handle
`import matplotlib` -> `matplotlib.use()` and `import matplotlib as mpl` -> `mpl.use()`; I added the second alias because it is commonly used. But one could also use `import matplotlib as m`, and the resulting `m.use()` will go unnoticed. (I added a note about that to the hook config documentation).

The `from matplotlib import use` -> `use()` case is also covered by `any_alias("matplotlib.use")`. But it will result in a false positive for the following code snipptet:

```python
import matplotlib

def use(name):
    # do something completely unrelated to matplotlib, with passed 
    # name being something different from mpl backend name

matplotlib.use('TkAgg')
use('something_unrelated_to_mpl')
```

The false positive should end up generating an invalid hidden import (so a warning about it not existing), but still, it would be useful if we could verify that the matched symbol is indeed bound to what we expect it to be (i.e., `matplotlib.use` if `from matplotlib import use` was used).

But this does indeed fall beyond the ability of bytecode scanning (in the live code, we could probably extract that information via `inspect`).